### PR TITLE
Repo rule with gazelle support

### PR DESCRIPTION
**Pull request by srikrsna-buf**
Closes TCN-164, TCN-165, and TCN-166.

Gazelle support:
- `gazelle` will resolve proto files that are downloaded using this rule.
- `gazelle` can be used to import from `buf.work.yaml` and `buf.yaml` (provided the `buf.lock` files are present)

### CODE REVIEW

File: buf/internal/repo.bzl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        "-repo_root",
+        ctx.path(""),
+    ]
+    res = ctx.execute(cmd, quiet = False)
+    if res.return_code != 0:
+        fail("failed with code: {}, error: {}".format(res.return_code, res.stderr))
+
+    return update_attrs(ctx.attr, ["deps"], {})
+
+buf_dependencies = repository_rule(
```
Comment (srikrsna-buf) [inline buf/internal/repo.bzl:67]: I am not fan of the name `buf_dependencies`. Any suggestions?

File: examples/gazelle/buf_deps.bzl [hunk 2]
```diff
@@ -0,0 +1,11 @@
+load("@rules_buf//buf:defs.bzl", "buf_dependencies")
+
+def buf_deps():
+    buf_dependencies(
+        name = "buf_deps_barapis",
+        deps = ["buf.build/acme/paymentapis:6e230f46113f498392c82d12b1a07b70", "buf.build/acme/petapis:84a33a06f0954823a6f2a089fb1bb82e", "buf.build/envoyproxy/protoc-gen-validate:dc09a417d27241f7b069feae2cd74a0e", "buf.build/googleapis/googleapis:84c3cad756d2435982d9e3b72680fa96"],
```
Comment (bufdev) [inline examples/gazelle/buf_deps.bzl:6]: These should probably be generated to separate lines

**Review by bufdev — COMMENTED** [review]
I'm not qualified to review the starlark, but I'm less concerned about that, what I do want to see:

- More docs on how to use this
- Development documentation, specifically around the Golang code. Imagine you're another developer coming in and needing to edit this or understand what is going on - where do I start? There looks to be cross-linking? What's the entry point, where can I read about this? The code needs a lot of documentation I believe.

File: examples/gazelle/buf_deps.bzl (hunk 2 above)
Comment (srikrsna-buf) [reply examples/gazelle/buf_deps.bzl:6]: Yes, I would like that too but we don't have any control over this, we just provide the rule, gazelle does the printing and formatting.

File: buf/internal/repo.bzl [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    if len(module_parts) != 3:
+        return False
+
+    return True
+
+def _buf_dependencies_impl(ctx):
+    buf = ctx.path(Label("@{}//:buf{}".format(ctx.attr.toolchain_repo, _executable_extension(ctx))))
+
+    for pin in ctx.attr.deps:
+        if not _valid_pin(pin):
```
[the remark this answers is not in the document]
Comment (amckinney) [reply buf/internal/repo.bzl:33]: I agree with offloading the responsibility to `gazelle` - moves commit resolution outside of the build step. We can always revisit that later if we need to, but this seems like the right place to start.

File: buf/internal/repo.bzl (hunk 1 above)
Comment (amckinney) [reply buf/internal/repo.bzl:67]: For consistency with the `deps` key in the configuration, we could shorten it to `buf_deps`. We'll eventually have the concept of plugin dependencies (via `buf.plugin.yaml`), so we might want this to be `buf_module_deps`, but that's a little busy too.

We could always reserve `buf_deps` for this rule, then roll out a `buf_plugin_deps` if we ever needed it. I'm not even sure that we'll ever need a rule like `buf_plugin_deps` that for `buf.plugin.yaml` definitions in Bazel, so it's probably a little overkill anyway.

TL;DR I'd probably just name this `buf_deps`.

Comment (amckinney) [reply buf/internal/repo.bzl:67]: Ah I spoke too earlier - I see we already have a `buf_deps` rule used elsewhere. If it's a matter of which one is more user facing, we should make the user-facing one `buf_deps`, but it looks like that's already the case.

Comment (srikrsna-buf) [reply buf/internal/repo.bzl:67]: How about `buf_repository` like `git_repository`, and `local_repository` and renaming the `deps` attribute to `modules`?

Comment (bufdev) [reply buf/internal/repo.bzl:67]: `buf_repository` seems on face value to be inconsistent with our naming, same with `deps` to `modules`. Having `buf_deps` and `buf_dependencies` worries me too. I'd finish the docs first, and then implement what makes most sense from a consistency perspective.

**Comment by srikrsna-buf** [comment]
> More docs on how to use this

I think we should finalise this in the [docs PR](https://github.com/bufbuild/docs.buf.build/pull/160), then I will replicate that in the repo to keep it consistent.

>  Development documentation

I agree that the gazelle extension needs documentation not just for the cross resolver but overall. The only available resources are [Architecture of Gazelle](https://github.com/bazelbuild/bazel-gazelle/blob/master/Design.rst), [Extending Gazelle](https://github.com/bazelbuild/bazel-gazelle/blob/master/extend.md) and of course the [godoc](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle#section-readme). I'll add these links in a developer readme in a separate PR along with a general description of the package structure.

File: gazelle/buf/buf_lock.go [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	Version string `yaml:"version,omitempty" json:"version,omitempty"`
+	Deps    []struct {
+		Remote     string `yaml:"remote,omitempty" json:"remote,omitempty"`
+		Owner      string `yaml:"owner,omitempty" json:"owner,omitempty"`
+		Repository string `yaml:"repository,omitempty" json:"repository,omitempty"`
+		Commit     string `yaml:"commit,omitempty" json:"commit,omitempty"`
+	} `yaml:"deps,omitempty" json:"deps,omitempty"`
+}
+
+func readBufLock(file string) (*bufLock, error) {
```
Comment (robbertvanginkel) [inline gazelle/buf/buf_lock.go:15]: Nit: `readBufWork` and `readBufLock` seem almost identical and are both only used in one occasion. I'd remove these helpers and just inline the `os.ReadFile`/`parseJsonOrYaml` rather than adding the layer of indirection.

File: gazelle/buf/buf_lock.go [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package buf
+
+import "os"
+
+type bufLock struct {
```
Comment (robbertvanginkel) [inline gazelle/buf/buf_lock.go:5]: I think we should have some comment to what code to bufbuild/buf this should be kept in sync to (assuming sharing structs with bufbuild/buf is out of the question).

File: gazelle/buf/repositories.bzl [hunk 6]
```diff
@@ -1,19 +1,28 @@
...
+        sum = "h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=",
+        version = "v3.0.1",
+    )
+    http_archive(
+        name = "com_github_bazelbuild_buildtools",
+        sha256 = "ae34c344514e08c23e90da0e2d6cb700fcd28e80c02e23e4d5715dddcb42f7b3",
+        strip_prefix = "buildtools-4.2.2",
+        urls = [
+            "https://github.com/bazelbuild/buildtools/archive/refs/tags/4.2.2.tar.gz",
+        ],
```
Comment (robbertvanginkel) [inline gazelle/buf/repositories.bzl:27]: Why do we use `go_repository` for the other golang dependencies but http_archive for this one? Are there special update instructions (other than the gazelle update-repos for the `go_repository` ones)?

File: buf/internal/repo.bzl [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+        extension = ".exe"
+    return extension
+
+def _valid_pin(pin):
+    module_commit = pin.split(":")
+    if len(module_commit) != 2:
+        return False
+
+    if not module_commit[1].isalnum():
+        return False
```
[the remark this answers is not in the document]
Comment (srikrsna-buf) [reply buf/internal/repo.bzl:21]: Based on our suggestion [here](https://docs.buf.build/how-to/iterate-on-modules#edit-and-push). We can't stop the from not using tags, we can only stop them from using some forms of tags. So I'll just remove the check here.

File: gazelle/buf/repositories.bzl (hunk 6 above)
Comment (srikrsna-buf) [reply gazelle/buf/repositories.bzl:27]: I already had that as a `http_archive` and gazelle didn't generate a `go_repository` rule, I thought it was a special case. Now that I remove the build tools and run it again I see it pop up. I'll change this to `go_repository`

File: examples/gazelle/buf_deps.bzl (hunk 2 above)
Comment (srikrsna-buf) [reply examples/gazelle/buf_deps.bzl]: Now fixed thanks to Robbert

File: buf/internal/repo.bzl (hunk 1 above)
Comment (srikrsna-buf) [reply buf/internal/repo.bzl:64]: So `deps` can't be used as an attr, cause things like aspects and bazel in general expects them to be labels, see https://github.com/bufbuild/rules_buf/pull/14#discussion_r898269203. I changed the attr to `modules`

Comment (robbertvanginkel) [reply buf/internal/repo.bzl:64]: The canonical example for using an auto updated macro to track deps is in https://github.com/bazelbuild/bazel-gazelle#update-repos, which works something like
```
# repositories.bzl
def go_repositories():
    # The go_repository rules below get updated by running:
    #   $ gazelle update-repos -from_file=go.work -to_macro=repositories.bzl%go_repositories
    go_repository(...)
    go_repository(...)

# WORKSPACE
load("//:repositories.bzl", "go_repositories")

go_repositories() # this loads all the go_repository rules here into the workspace
```

When loading dependencies provided by external rulesets, the convention is to have something like a `_dependencies` in a `deps.bzl`:
```
load("@io_bazel_rules_go//go:deps.bzl", "go_rules_dependencies")
go_rules_dependencies()
load("@bazel_gazelle//:deps.bzl", "gazelle_dependencies")
gazelle_dependencies()
```
However, our usage of `buf_deps` in this PR is not a rule provided by rules_buf, but a macro name used for the current WORKSPACE's dependencies in one of the examples. As far as I know, there are no conventions for that other than maybe `repositories.bzl%go_repositories` mentioned in the gazelle documentation (usually others should not be importing your rules' development dependencies).

However, in our case the current `buf_dependencies` rule maps to downloading a list of repositories for a `buf.work`/`buf.lock`. This is unlike the more granular repo rule you'd see in bazel such as the `go_repository` rule which would only download an individual go module (and you'd generate many of those per go.mod, unlike the current 1 buf_dependencies per buf.lock/buf.work).

If I recall, the problem with doing a rule per buf module approach is with workspaces. Different workspace modules can have different versions of a common dependency pinned, so either we'd have to put versions into rule names or we couldn't replicate the cli build behavior.

--- 

Given the above, a `buf_dependencies` rule that has multiple `module` seems the most straightforward to replicate workspaces in bazel.

One last thing to note is that since bazel does put a premium on build reproducibility, dependency rules should generally have some digest in them for making sure the content received for a pointer is the same when the rule was committed as in the future. `go_repository` has `sum` for this, `http_archive` uses `sha256`. Ideally we'd have something like this for `buf_dependencies`, but with the sums being absent from buf.lock I don't think we can really do much here without first addressing that in the cli itself.

**Comment by robbertvanginkel** [comment]
> The only available resources are [Architecture of Gazelle](https://github.com/bazelbuild/bazel-gazelle/blob/master/Design.rst), [Extending Gazelle](https://github.com/bazelbuild/bazel-gazelle/blob/master/extend.md) and of course the [godoc](https://pkg.go.dev/github.com/bazelbuild/bazel-gazelle#section-readme). I'll add these links in a developer readme in a separate PR along with a general description of the package structure.

Maybe lets just add them to this PR? Should only be a small change to the toplevel readme

**Review by robbertvanginkel — APPROVED** [review]
